plasma_framework/contracts/src/exits/payment/PaymentInFlightExitable.sol

```diff
@@ -108,6 +128,73 @@ contract PaymentInFlightExitable is
         emit InFlightExitStarted(msg.sender, startExitData.inFlightTxHash);
     }
 
+    function challengeInFlightExitNotCanonical(ChallengeCanonicityArgs memory args) public payable {
+        // Check if there is an active in-flight exit from this transaction?
+        uint192 exitId = ExitId.getInFlightExitId(args.inFlightTx);
+        PaymentExitDataModel.InFlightExit storage ife = inFlightExits[exitId];
+        require(ife.exitStartTimestamp != 0, "In-fligh exit doesn't exists");
+
+        // Check that the exit is active and in period 1
+        verifyFirstPhaseNotOver(ife);
+        // Check if exit's input was spent via MVP exit
+        verifyInputNotSpent(ife);
+
+        // Check that two transactions are not the same
+        require(
+            keccak256(args.inFlightTx) != keccak256(args.competingTx),
+            "The competitor transaction is the same as transaction in-flight"
+        );
+
+        // Check that both transactions share the same input
+        PaymentTransactionModel.Transaction memory inFlight = PaymentTransactionModel.decode(args.inFlightTx);
+        PaymentTransactionModel.Transaction memory competitor = PaymentTransactionModel.decode(args.competingTx);
+        require(
+            inFlight.inputs[args.inFlightTxInputIndex] == competitor.inputs[args.competingTxInputIndex],
+            "The competitor and transcation in-flight have to share input at given positions"
+        );
+
+        // Check that shared input owner signes competing transaction
+        require(
+            isSpendingConditionMet(
+                ife.inputs[args.inFlightTxInputIndex].outputGuard,
+                args.competingTxInputPos,
```

Comment:
This and the `outputId` should be some already checked `utxoPos` or `outputId`. In SE, both of them are checked while exit starts and then saved. Then challenge simply pass in the checked data from storage. Not sure what's best to do here though.